File: gfapy/oriented_line.py
import gfapy
import re

class OrientedLine:
  """A line plus an orientation.

  The line can be an instance of `~gfapy.line.line.Line` or a string (line
  identifier). The orientation is a string, either ``'+'`` or ``'-'``.
  Methods not defined in this class are delegated to the line element.

  Parameters:
    value (str, list, OrientedLine) : a line identifier with a 1-letter
      orientation suffix + or -, or a list of two elements (identifier
      or line instance and orientation string), or an OrientedLine instance

  Returns:
    OrientedLine: if value is an OrientedLine, then
      it is returned; if it is a string, then an OrientedLine where line
      is a string (the string without the last char, which is the orientation);
      if it is a list, then an OrientedLine where line is the first element,
      orientation the second
  """

  def __new__(cls, *args):
    if isinstance(args[0], OrientedLine):
      return args[0]
    else:
      new_instance = object.__new__(cls)
      return new_instance

  def __init__(self, *args):
    if len(args) == 1:
      if isinstance(args[0], OrientedLine):
        return
      elif isinstance(args[0], str):
        self.__line = args[0][0:-1]
        self.__orient = args[0][-1]
      elif isinstance(args[0], list):
        self.__line = args[0][0]
        self.__orient = args[0][1]
      else:
        raise gfapy.ArgumentError("Cannot create an OrientedLine"+
            " instance from an object of type {}".format(type(args[0])))
    elif len(args) == 2:
      self.__line = args[0]
      self.__orient = args[1]
    else:
      raise gfapy.ArgumentError("Wrong number of arguments for OrientedLine()")
    self.__editable = True

  @property
  def line(self):
    """The line.

    Returns:
      str or `~gfapy.line.line.Line`
    """
    return self.__line

  @property
  def orient(self):
    """The orientation.

    Returns:
      str : '+' or '-'
    """
    return self.__orient

  @line.setter
  def line(self, line):
    if self.__editable:
      self.__line = line
    else:
      raise gfapy.RuntimeError(
          "gfapy.OrientedLine instance cannot be edited ({})".format(self))

  @orient.setter
  def orient(self, orient):
    if self.__editable:
      self.__orient = orient
    else:
      raise gfapy.RuntimeError(
          "gfapy.OrientedLine instance cannot be edited ({})".format(self))

  @property
  def name(self):
    """The name of the line.

    Returns:
      str : if line is a string, then line; if it is a line instance,
            then line.name
    """
    if isinstance(self.__line, str):
      return self.__line
    else:
      return self.__line.name

  def validate(self):
    """Validate the content of the instance

    Raises:
      gfapy.error.ValueError: if the orientation is invalid
      gfapy.error.TypeError: if the line is not a string or a
        `gfapy.line.line.Line` instance
      gfapy.error.FormatError: if the line is a string which is not a valid
        line identifier, or it is a Line instance with an invalid name
    """
    self.__validate_line()
    self.__validate_orient()
    return None

  def inverted(self):
    """An oriented line with the same line element, but inverted orientation.

    Note:
      the inverted() method returns an OrientedLine with inverted orientation;
      the invert() method inverts the orientation in place (and returns None)
    """
    return OrientedLine(self.line, gfapy.invert(self.orient))

  def invert(self):
    """Invert the orientation of the OrientedLine instance.

    Note:
      the inverted() method returns an OrientedLine with inverted orientation;
      the invert() method inverts the orientation in place (and returns None)
    """
    self.orient = gfapy.invert(self.orient)

  def __str__(self):
    if self.name:
      return "{}{}".format(self.name, self.orient)
    else:
      return "({}){}".format(str(self.line), self.orient)

  def __repr__(self):
    return "gfapy.OrientedLine({},{})".format(repr(self.line),repr(self.orient))

  def __eq__(self, other):
    if isinstance(other, OrientedLine):
      pass
    elif isinstance(other, list):
      other = OrientedLine(other)
    elif isinstance(other, str):
      other = OrientedLine(other)
    else:
      return False
    return (self.name == other.name) and (self.orient == other.orient)

  # Delegate methods to the line
  def __getattr__(self, name):
    return getattr(self.__line, name)

  def _block(self):
    self.__editable = False

  def _unblock(self):
    self.__editable = True

  def __validate_orient(self):
    if not self.orient in ["+", "-"]:
      raise gfapy.ValueError("Invalid orientation ({})".format(self.orient))

  def __validate_line(self):
    if isinstance(self.line, gfapy.Line):
      string = self.line.name
    elif isinstance(self.line, str):
      string = self.line
    else:
      raise gfapy.TypeError(
        "Invalid class ({}) for line reference ({})"
        .format(self.line.__class__, self.line))
    if not re.match(r"^[!-~]+$", string):
      raise gfapy.FormatError(
      "{} is not a valid GFA identifier\n".format(repr(string))+
      "(it contains spaces or non-printable characters)")

File: lib/rgfa/oriented_line.rb
require_relative "error"

# A line or line identifier plus an orientation.
#
# @tested_in unit_oriented_line
class RGFA::OrientedLine

  # Create an OrientedLine instance
  # @param line [Symbol, RGFA::Line] either a line identifier (Symbol)
  #   or a reference to a line (RGFA::Line subclass instance)
  # @param orient [:+,:-] a symbol defining the orientation
  def initialize(line, orient)
    @line = line
    @orient = orient.to_sym
    @editable = true
  end

  # Get the line
  # @return [RGFA::Line, Symbol] the line identifier or reference
  def line
    @line
  end

  # Get the orientation symbol
  # @return [:+,:-] the orientation symbol
  def orient
    @orient
  end

  # Set the line
  # @param line [Symbol, RGFA::Line] set the line instance
  # @raise [RGFA::RuntimeError] if the line instance is not editable
  # @return [Symbol, RGFA::Line] the line
  def line=(line)
    if @editable
      @line = line
    else
      raise RGFA::RuntimeError,
        "RGFA::OrientedLine instance cannot be edited (#{self})"
    end
  end

  # Set the orientation
  # @param orient [:+,:-] set the orientation symbol
  # @raise [RGFA::RuntimeError] if the line instance is not editable
  # @return [:+,:-] the orientation
  def orient=(orient)
    if @editable
      @orient = orient
    else
      raise RGFA::RuntimeError,
        "RGFA::OrientedLine instance cannot be edited (#{self})"
    end
    return @orient
  end

  # @return [Symbol] the line name
  def name
    @line.to_sym
  end

  # Validate the instance
  # @raise [RGFA::ValueError] if the orientation symbol is not +:++ or +:-+
  # @raise [RGFA::TypeError] if the line is not a string, symbol
  #   or reference to a line (RGFA::Line)
  # @raise [RGFA::FormatError] if the line is a string or symbol and
  #   it contains spacing or non-printable characters
  # @return [void]
  def validate
    validate_line
    validate_orient
    return nil
  end

  # Create an oriented line instance, with the inverted orientation
  # @return [RGFA::OrientedLine] same line, inverted orientation
  def invert
    RGFA::OrientedLine.new(@line, @orient.invert)
  end

  # Compute the string representation of the oriented line
  # @return [String] line name and orientation
  def to_s
    "#{name}#{orient}"
  end

  # Compare the segment names and orientations of two instances
  #
  # @param [RGFA::OrientedLine,Array] other the other instance
  # @return [Boolean]
  def ==(other)
    case other
    when RGFA::OrientedLine
    when Array
      other = other.to_oriented_line
    when String, Symbol
      other = other.to_s.to_oriented_line
    else
      return false
    end
    (self.name == other.name) and
      (self.orient == other.orient)
  end

  # Return self (for compatibility with to_oriented_line of other classes)
  # @return [RGFA::OrientedLine] self
  def to_oriented_line
    self
  end

  # @api private
  module API_PRIVATE

    def block
      @editable = false
    end

    def unblock
      @editable = true
    end
  end

  include API_PRIVATE

  private

  # Delegate methods to the line
  def method_missing(meth, *args, &block)
    @line.send meth, *args, &block
  end

  def validate_orient
    if ![:+,:-].include?(@orient)
      raise RGFA::ValueError,
        "Invalid orientation (#{@orient})"
    end
  end

  def validate_line
    case @line
    when RGFA::Line
      string = @line.name
    when Symbol, String
      string = @line
    else
      raise RGFA::TypeError,
        "Invalid class (#{@line.class}) for line reference (#{@line})"
    end
    if string !~ /^[!-~]+$/
      raise RGFA::FormatError,
      "#{string.inspect} is not a valid GFA identifier\n"+
      "(it contains spaces or non-printable characters)"
    end
  end

end

class String
  # Create an oriented line instance from a string, which contains
  # a GFA identifier followed by + or -
  # @return [RGFA::OrientedLine]
  # @tested_in unit_oriented_line
  def to_oriented_line
    RGFA::OrientedLine.new(self[0..-2].to_sym, self[-1].to_sym)
  end
end

class Array
  # Create an oriented line instance from an array, which contains
  # a GFA identifier as first element, and a :+ or :- as second
  # @return [RGFA::OrientedLine]
  # @tested_in unit_oriented_line
  def to_oriented_line
    RGFA::OrientedLine.new(self[0], self[1].to_sym)
  end
end

# Shortcut to create new objects
# using a OL[] syntax
module OL
  # Create an oriented line instance
  # (shortcut for RGFA::OrientedLine.new)
  # @param line [RGFA::Line,Symbol] the line
  # @param orient [:+,:-] the orientation symbol
  # @return [RGFA::OrientedLine]
  def [](line,orient)
    RGFA::OrientedLine.new(line,orient)
  end
  module_function :[]
end
Kernel.extend(OL)
